### PR TITLE
Fix GPKG/SQLITE comments in query layers

### LINKED ISSUE #45754: Query returns empty layer if the first row is empty or commented out
### What is the bug or the crash?

if I write a query and don't leave the first row blank, the query works fine and returns correct output.
adding the first row and leaving it blank or with a comment, generates an empty layer.

https://user-images.githubusercontent.com/7631137/139591217-bc4f0f86-7086-46a7-b5ce-eda229226d7b.mp4

### Steps to reproduce the issue

1. create any query using the browser panel
2. leave the first line blank or comment it out
3. execution generates correct output
4. but if I add the layer in QGIS, the layer is empty

## OSGeo4W64 v2 QGIS 3.22.0
QGIS version 3.22.0-Białowieża
QGIS code revision d902269
Qt version 5.15.2
Python version 3.9.5
GDAL/OGR version 3.3.2
PROJ version 8.1.1
EPSG Registry database version v10.028 (2021-07-07)
GEOS version 3.10.0-CAPI-1.16.0
SQLite version 3.35.2
PDAL version 2.3.0
PostgreSQL client version 13.0
SpatiaLite version 5.0.1
QWT version 6.1.3
QScintilla2 version 2.11.5
OS version Windows 10 Version 2009

Active Python plugins db_manager 0.1.20
grassprovider 2.12.99
MetaSearch 0.3.5
processing 2.12.99
sagaprovider 2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

same problem also in the master

### PR DESCRIPTION
**Pull request by elpaso**
Fix #45754

### CODE REVIEW

**Comment by elpaso** [comment]
unrelated test failure

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 1]
```diff
@@ -1300,6 +1300,30 @@ QString QgsOgrProviderUtils::quotedValue( const QVariant &value )
...
+  lines.erase( std::remove_if( lines.begin(), lines.end(), []( const QString & line )
+  {
+    return line.startsWith( QStringLiteral( "--" ) );
+  } ), lines.end() );
+  for ( auto &line : lines )
+  {
+    bool inLiteral {false};
+    for ( int i = 0; i < line.length(); ++i )
+    {
+      if ( line[i] == QChar( '\'' ) && ( i == 0 || line[i - 1] != QChar( '\\' ) ) )
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:1314]: What would happen if a column name escaped in ".." had a "--" in it?

Comment (elpaso) [reply src/core/providers/ogr/qgsogrproviderutils.cpp]: fixed

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 2]
```diff
@@ -1300,6 +1300,32 @@ QString QgsOgrProviderUtils::quotedValue( const QVariant &value )
...
+  {
+    return line.startsWith( QStringLiteral( "--" ) );
+  } ), lines.end() );
+  for ( auto &line : lines )
+  {
+    bool inLiteral {false};
+    QChar literalChar { ' ' };
+    for ( int i = 0; i < line.length(); ++i )
+    {
+      if ( ( ( line[i] == QChar( '\'' ) || line[i] == QChar( '"' ) ) && ( i == 0 || line[i - 1] != QChar( '\\' ) ) ) && ( line[i] != literalChar ) )
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:1315]: Hate to be annoying, but won't this still be broken for some cases? e.g.

    select "my--thing's" from something where "thing" = 'my "things -- all'

(Of course, anyone who has a database like that needs a good punch in the face)

Comment (elpaso) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:1315]: Actually, that case was already handled correctly, I added it to the test: https://github.com/qgis/QGIS/pull/45797/commits/68d69cb7d1a243276213a58c8f625d8bf3680159#diff-56354e2446fe2cb6d1ee92d4e984091172e964e90f3be4d3d42276e033c4986eR2370

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**